public/src/utils.js

```diff
@@ -469,6 +470,48 @@
 			return utils.extensionMimeTypeMap[extension] || '*';
 		},
 
+		isPromise: function (object) {
+			// https://stackoverflow.com/questions/27746304/how-do-i-tell-if-an-object-is-a-promise#comment97339131_27746324
+			return object && typeof object.then === 'function';
+		},
+
+		// todo: use js template literals in 2030
+		// todo: use js yield* in 2040
+		callable: function (func, options) {
+			options = options || {};
+			options.times = parseInt(options.times, 10) || null;
+			options.timeout = parseInt(options.timeout, 10) || null;
+			options.onTimeout = typeof options.onTimeout === 'function' ? options.onTimeout : function () {};
+			var called = 0;
+			var timeoutId;
+			var timedout = false;
+			if (options.timeout) {
+				timeoutId = setTimeout(function () {
+					timedout = true;
+					clearTimeout(timeoutId);
+					options.onTimeout();
+				}, options.timeout);
+			}
+			return function () {
+				clearTimeout(timeoutId);
+				var repeated = options.times && options.times > 0 && called >= options.times;
+				var warningPrepend;
+				if (repeated || timedout) {
+					warningPrepend = 'Function named: \'' + (func.name || func.constructor.name) + '\', arguments:[\'' + Array.from(arguments).join('\', \'') + '\']';
+				}
+				if (timedout) {
```

Comment:
Nest these two if statements inside the one above to avoid the separate `var warningPrepend;` above